### DISCUSSION TITLE
creating card token for testing integration with backend

### DISCUSSION THREAD

**Discussion by ihojmanb** [mercadopago/sdk-js#101] [Q&A] [answered]
Is there a way to create card tokens from the backend? Currently, the card token is generated through the clientside SDK with `mercadopago.fields.createCardToken(...)`. This is a bit painful when trying to test my backend, as I have to go all the way from filling in the payment form by hand, generate the token, and make a POST request to receive that token into my backend.

Is there a smart way to generate card tokens for testing purposes or how can I import the clientside SDK in my backend?

**Comment by icaldana** [comment, marked as answer]
Hello @ihojmanb! Let me try to understand your pain.

1. Are you using [backend SDK](https://www.mercadopago.cl/developers/es/docs/checkout-pro/requirements#editor_2:~:text=Instala%20la%20SDK%20de%20Mercado%20Pago)? 
2. What kind of tests are you creating? At the first glance it seems you're running E2E tests, and if yes, it does make sense to include both FE and BE in the process. On the other hand, if the tests are unitary or even manual, couldn't you mock the token and save it as one environment variable?

**Reply by icaldana** [reply]
Normally, real card tokens are valid only once, so it depends on what kind of tests are you running 🤔

**Reply by ihojmanb** [reply]
Hi @icaldana! sorry for not being that specific. Here's how it goes:

1. I'm integrating with the [Checkout API V2 via Core Methods](https://www.mercadopago.cl/developers/en/docs/checkout-api-v2/receiving-payment-by-card-core-methods). So whenever I need to make a payment, I do something like this:
```js
var payment_data = {
  transaction_amount: Number(req.body.transactionAmount),
  token: req.body.token,
  description: req.body.description,
  installments: Number(req.body.installments),
  payment_method_id: req.body.paymentMethodId,
  issuer_id: req.body.issuer,
  payer: {
    email: req.body.email,
  }
};

mercadopago.payment.save(payment_data)
```
What's bothering me here is the `token` value, since I (think) have to generate a valid token for the [available test cards](https://www.mercadopago.cl/developers/en/docs/checkout-api-v2/testing).
2. I'm testing it manually first, then I'm going with unit/E2E tests. How can I mock a card token to be a valid one?

I think all comes down to: how can I generate a valid token for the test cards available without passing by the front end?

**Reply by icaldana** [reply]
Thank you for clarifying it 😄

Unfortunately, this is part of [PCI](https://en.wikipedia.org/wiki/Payment_Card_Industry_Data_Security_Standard) [standards](https://www.pcisecuritystandards.org/), thus, this is the expected behavior for these tests E2E/manual for security reasons. 

In a nutshell: in order to test your custom logic you could mock the token, but when testing the SDK-js integration, it is required to pass by the front-end.

**Reply by ihojmanb** [reply]
Do you have any advice on how to mock the token? I tried using one shown in the documentation, but received `'Card Token not found'` error (code `2006`)

thanks for your help Ismael! 😄

**Reply by icaldana** [reply]
Just clarifying the tokens are designed to be single-use, then mocking makes sense only when you don't need to actually call `mercadopago.payment.save`. You could store a token example as environment variable https://nodejs.dev/en/learn/how-to-read-environment-variables-from-nodejs/

**Reply by ihojmanb** [reply]
thanks, gonna check it out!
